src/oop/programs_with_multiple_classes/hard/q4/MainTest.java

```diff
@@ -0,0 +1,150 @@
+
+
+    @BeforeEach
+    public void Setup() {
+        String bookClassString = "oop.programs_with_multiple_classes.hard.q4." + bookUc;
+        String bookcaseClassString = "oop.programs_with_multiple_classes.hard.q4." + bookcaseUc;
+        book = new ObjectTest(bookClassString);
+        bookcase = new ObjectTest(bookcaseClassString);
+    }
+
+    /**
+     * Book tests
+     **/
+    @Test
+    public void bookClassHasCorrectFields() {
+        String missingFieldMessage = "Your " + bookLc + " class is missing a required field.";
+        String incorrectFieldModifierMessage = "A field in your " + bookLc + " class has an the wrong visibility modifier.";
+        assertTrue(book.hasField(year, int.class), missingFieldMessage);
+        assertTrue(book.hasField(type, String.class), missingFieldMessage);
+        assertTrue(book.hasModifier(year, "private"), incorrectFieldModifierMessage);
+        assertTrue(book.hasModifier(type, "private"), incorrectFieldModifierMessage);
+    }
+
+    @Test
+    public void bookClassHasCorrectConstructors() {
+        String missingConstructorMessage = "Your " + bookLc + " class does not have a required constructor.";
+        String incorrectConstructorModifierMessage =
+                "One of the constructors in your " + bookLc + " class does not have the correct visibility modifier.";
+        Class<?>[] constructorClasses = new Class[]{
+                int.class,
+                String.class
+        };
+        assertTrue(book.hasConstructor(constructorClasses), missingConstructorMessage);
+        assertTrue(book.hasModifier(constructorClasses, "public"), incorrectConstructorModifierMessage);
+    }
+
+    private static Stream<Arguments> bookInputProvider() {
+        return Stream.of(
+                Arguments.of(1, "Hardcover"),
+                Arguments.of(1, "Softcover"),
+                Arguments.of(3, "Paperback"),
+                Arguments.of(4, "Softback"),
+                Arguments.of(0, "Hardcover"));
+    }
+
+    @ParameterizedTest
+    @MethodSource("bookInputProvider")
+    public void bookConstructorInitializesValuesCorrectly(int year, String type) throws Throwable {
+        Object[][] arguments = new Object[][]{
+                {year, int.class},
+                {type, String.class}
+        };
+        Object bookInstance = book.createInstance(arguments);
+        _assertEquals(year, book.getFieldValue(bookInstance, this.year),
+                "Your " + book + " constructor does not correctly initialize the " + this.year + " field.");
+        _assertEquals(type, book.getFieldValue(bookInstance, this.type),
+                "Your " + book + " constructor does not correctly initialize the " + this.type + " field.");
+    }
+
+    @ParameterizedTest
+    @MethodSource("bookInputProvider")
+    public void correctHowManyYearsMethod(int year, String type) throws Throwable {
+        Object[][] arguments = new Object[][]{
```

Comment:
Also, for methods we decided to add a check if they exist before calling them. That way we can check things like the visibility modifier separately from the actual method definition, and we can also provide a custom error message based on which part is wrong instead of the default message that it could not be found.